nav2_map_server/src/map_saver.cpp

```diff
@@ -30,221 +31,173 @@
+MapSaver::on_error(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_FATAL(get_logger(), "Lifecycle node entered error state");
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+MapSaver::on_shutdown(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Shutting down");
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+bool MapSaver::saveMapTopicToFile(std::string & map_topic, SaveParameters & save_parameters)
+{
+  RCLCPP_INFO(get_logger(), "Saving map from %s topic to %s file",
+    map_topic.c_str(), save_parameters.map_file_name.c_str());
+
+  try {
+    // Reset map receiving indicator
+    got_map_msg_ = false;
+
+    // Correct map_topic if necessary
+    if (map_topic == "") {
+      map_topic = "map";
       RCLCPP_WARN(
-        get_logger(), "Requested image format '%s' is not one of the recommended formats: %s",
-        image_format.c_str(), ss.str().c_str());
+        get_logger(), "Map topic unspecified. Map messages will be read from %s topic",
+        map_topic.c_str());
     }
-    const std::string FALLBACK_FORMAT = "png";
-
-    try {
-      Magick::CoderInfo info(image_format);
-      if (!info.isWritable()) {
-        RCLCPP_WARN(
-          get_logger(), "Format '%s' is not writable. Using '%s' instead",
-          image_format.c_str(), FALLBACK_FORMAT.c_str());
-        image_format = FALLBACK_FORMAT;
-      }
-    } catch (Magick::ErrorOption & e) {
-      RCLCPP_WARN(
-        get_logger(), "Format '%s' is not usable. Using '%s' instead:\n%s",
-        image_format.c_str(), FALLBACK_FORMAT.c_str(), e.what());
-      image_format = FALLBACK_FORMAT;
+
+    // Set default for MapSaver node thresholds parameters
+    if (save_parameters.free_thresh == 0) {
+      RCLCPP_WARN(get_logger(),
+        "Free threshold unspecified. Setting it to default value: %i", free_thresh_default_);
+      save_parameters.free_thresh = free_thresh_default_;
     }
-    if (
-      map_mode == MapMode::Scale &&
-      (image_format == "pgm" || image_format == "jpg" || image_format == "jpeg"))
-    {
-      RCLCPP_WARN(
-        get_logger(),
-        "Map mode 'scale' requires transparency, but format '%s' does not support it. Consider "
-        "switching image format to 'png'.",
-        image_format.c_str());
+    if (save_parameters.occupied_thresh == 0) {
+      RCLCPP_WARN(get_logger(),
+        "Occupied threshold unspecified. Setting it to default value: %i", occupied_thresh_default_);
+      save_parameters.occupied_thresh = occupied_thresh_default_;
     }
 
-    RCLCPP_INFO(get_logger(), "Waiting for the map");
-    map_sub_ = create_subscription<nav_msgs::msg::OccupancyGrid>(
-      "map", rclcpp::SystemDefaultsQoS(),
+    // Add new subscription for incoming map topic
+    auto map_sub = map_listener_->create_subscription<nav_msgs::msg::OccupancyGrid>(
+      map_topic, rclcpp::SystemDefaultsQoS(),
       std::bind(&MapSaver::mapCallback, this, std::placeholders::_1));
```

Comment:
`mapCallback` should be a local lambda. This lets you remove the msg_ and got_map_msg_ class variables and just be in the local scope (since its just setting a trivial flag and copying some data).

Reply:
Done: https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-99885e6562ca50edd8cc80db6362d317R158